### PR TITLE
Build macOS installer

### PR DESCRIPTION
**Pull request by cosmo0920**
I noticed that macOS does not have linkable OpenSSL library by default.
So, we should also bundle OpenSSL library on macOS installer.

### CODE REVIEW

**Comment by cosmo0920** [comment]
スクリーンショット 2020-09-01 10 31 21
I've got built almost same appearance td-agent dmg.

File: td-agent/Rakefile [hunk 1]
```diff
@@ -544,6 +614,70 @@ class BuildTask
...
+    mkdir_p(openssldir)
+    File.open(File.join(openssldir, "cert.pem"), 'w', 0644) do |f|
+      f.write(valid_certificates.join("\n") << "\n")
+    end
+  end
+
+  def build_openssl
+    tarball = @download_task.file_openssl_source
+    source_dir = tarball.sub(/\.tar\.gz$/, '')
+
```
Comment (kenhys) [inline td-agent/Rakefile:649]: File.basename(tarball, ".tar.gz") ?

File: td-agent/Rakefile [hunk 2]
```diff
@@ -544,6 +614,70 @@ class BuildTask
...
+      "darwin64-x86_64-cc",
+    ]
+
+    cd(source_dir) do
+      # Currently, macOS only.
+      sh("perl", "./Configure", *(configure_opts + arch_opts))
+      sh("make", "depend")
+      sh("make")
+      # For `gem install` Ruby.
+      sh("make", "install")
```
Comment (kenhys) [inline td-agent/Rakefile:676]: could you explain a bit futhermore?

File: td-agent/Rakefile [hunk 3]
```diff
@@ -941,6 +1083,208 @@ EOS
...
+       "create", "-ov",
+       "-srcfolder", "dmg",
+       "-format", "UDRW",
+       "-volname", @volume_name,
+       @dmg_temp_name)
+  end
+
+  def get_attached_rw_dmg
+    @device = `hdiutil attach -readwrite -noverify -noautoopen #{@dmg_temp_name} | egrep '^/dev/' | sed 1q | awk '{print $1}'`.strip
+    sleep 5
```
Comment (kenhys) [inline td-agent/Rakefile:1126]: Is this sleep  required?

File: td-agent/Rakefile [hunk 4]
```diff
@@ -494,6 +561,9 @@ class BuildTask
 
       desc "Create configuration files for Windows"
       task :msi_config => [:td_agent_config, :wix_config, :win_batch_files]
+
+      desc "Create configuration files for macOS"
+      task :pkg_config => [:td_agent_config, :pkgbuild_scripts, :pkgbuild_config, :bin_scripts, :launchctl_config]
```
Comment (kenhys) [inline td-agent/Rakefile:566]: It may be better to use pkgbuild_config or something. (dmg_config? to avoid confusing with `pkg-config` command)

**Review by kenhys — CHANGES_REQUESTED** [review]
I've added some comments.

File: td-agent/Rakefile (hunk 3 above)
Comment (cosmo0920) [reply td-agent/Rakefile:1126]: macOS Finder works a but slower than command line script execution...

File: td-agent/Rakefile (hunk 4 above)
Comment (cosmo0920) [reply td-agent/Rakefile:566]: dmg_config would be better for me.

File: td-agent/Rakefile (hunk 2 above)
Comment (cosmo0920) [reply td-agent/Rakefile:676]: Ruby which is installed in /opt/td-agent/bin needs to openssl library with the same way.
Same reason for make install execution on https://github.com/fluent-plugins-nursery/td-agent-builder/blob/54366045e988df59ca1e5af7d8822bbb79b72b7d/td-agent/Rakefile#L703.

File: td-agent/Rakefile (hunk 3 above)
Comment (kenhys) [reply td-agent/Rakefile:1126]: Got it. Thanks!

File: td-agent/Rakefile (hunk 2 above)
Comment (kenhys) [reply td-agent/Rakefile:676]: Got it. Thanks.

File: td-agent/Rakefile (hunk 1 above)
Comment (cosmo0920) [reply td-agent/Rakefile:649]: No. The suggested path should not work.
file_openssl_source should be included DOWNLOADS_DIR.
```
@file_openssl_source = File.join(DOWNLOADS_DIR, filename)
```

But your suggestion should drop DOWNLOADS_DIR from source_dir value.

Comment (kenhys) [reply td-agent/Rakefile:649]: Ah, I've overlooked it. :sob:

**Comment by cosmo0920** [comment]
I'd applied patches for @kenhys 's review.

File: Rakefile [hunk 5]
```diff
@@ -51,6 +51,7 @@ end
   ["msi:build",        "Build MSI package (alias for msi:dockerbuild)"],
   ["msi:selfbuild",    "Build MSI package without using Docker"],
   ["msi:dockerbuild",  "Build MSI package by Docker"],
+  ["mac:selfbuild",    "Build macOS package without using Docker"],
```
Comment (ashie) [inline Rakefile:54]: Should we use `dmg:selfbuild` since other tasks uses package names instead of OS name?

Comment (cosmo0920) [reply Rakefile:54]: Sure. Will do.

File: td-agent/Rakefile [hunk 6]
```diff
@@ -544,6 +615,74 @@ class BuildTask
...
+
+  def create_certs
+    keychains = [
+      "/System/Library/Keychains/SystemRootCertificates.keychain"
+    ]
+
+    cert_list, error, status = Open3.capture3("security find-certificate -a -p #{keychains.join(' ')}")
+    unless status.success?
+      puts(error: error)
+      exit 1
```
Comment (ashie) [inline td-agent/Rakefile:630]: We use `fail` method (alias of `raise`) for an error case.

Comment (cosmo0920) [reply td-agent/Rakefile:630]: Oh! I see.

**Comment by cosmo0920** [comment]
On w/o Homebrew macOS 10.15 environment:

### Dynamic Library Dependency

```
% otool -L /opt/td-agent/bin/ruby 
/opt/td-agent/bin/ruby:
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 59306.140.5)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1677.104.0)
	/opt/td-agent/lib/libruby.2.7.dylib (compatibility version 2.7.0, current version 2.7.1)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```

### Start td-agent service via Launchctl

```
% sudo launchctl load /Library/LaunchDaemons/td-agent.plist
Password:
```

### td-agent service status

```
% sudo launchctl list | grep td-agent
10238	0	td-agent
```

### td-agent log

```
% cat /var/log/td-agent/td-agent.log
2020-09-03 23:11:17 +0900 [info]: parsing config file is succeeded path="/etc/td-agent/td-agent.conf"
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-elasticsearch' version '4.1.1'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-kafka' version '0.14.1'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-prometheus' version '1.8.2'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-prometheus_pushgateway' version '0.0.2'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-record-modifier' version '2.1.0'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '2.3.0'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-s3' version '1.4.0'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-systemd' version '1.0.2'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-td' version '1.1.0'
2020-09-03 23:11:17 +0900 [info]: gem 'fluent-plugin-webhdfs' version '1.2.5'
2020-09-03 23:11:17 +0900 [info]: gem 'fluentd' version '1.11.2'
2020-09-03 23:11:17 +0900 [warn]: [output_td] Use different plugin for secondary. Check the plugin works with primary like secondary_file primary="Fluent::Plugin::TreasureDataLogOutput" secondary="Fluent::Plugin::FileOutput"
2020-09-03 23:11:18 +0900 [info]: using configuration file: <ROOT>
  <match td.*.*>
    @type tdlog
    @id output_td
    apikey xxxxxx
    auto_create_table 
    <buffer>
      @type "file"
      path "/var/log/td-agent/buffer/td"
    </buffer>
    <secondary>
      @type "file"
      path "/var/log/td-agent/failed_records"
      <buffer time>
        path /var/log/td-agent/failed_records
      </buffer>
    </secondary>
  </match>
  <match debug.**>
    @type stdout
    @id output_stdout
  </match>
  <source>
    @type forward
    @id input_forward
  </source>
  <source>
    @type http
    @id input_http
    port 8888
  </source>
  <source>
    @type debug_agent
    @id input_debug_agent
    bind "127.0.0.1"
    port 24230
  </source>
</ROOT>
2020-09-03 23:11:18 +0900 [info]: starting fluentd-1.11.2 pid=2886 ruby="2.7.1"
2020-09-03 23:11:18 +0900 [info]: spawn command to main:  cmdline=["/opt/td-agent/bin/ruby", "-Eascii-8bit:ascii-8bit", "/opt/td-agent/usr/sbin/td-agent", "--log", "/var/log/td-agent/td-agent.log", "--use-v1-config", "--under-supervisor"]
2020-09-03 23:11:21 +0900 [info]: adding match pattern="td.*.*" type="tdlog"
2020-09-03 23:11:21 +0900 [warn]: #0 [output_td] Use different plugin for secondary. Check the plugin works with primary like secondary_file primary="Fluent::Plugin::TreasureDataLogOutput" secondary="Fluent::Plugin::FileOutput"
2020-09-03 23:11:21 +0900 [info]: adding match pattern="debug.**" type="stdout"
2020-09-03 23:11:22 +0900 [info]: adding source type="forward"
2020-09-03 23:11:22 +0900 [info]: adding source type="http"
2020-09-03 23:11:22 +0900 [info]: adding source type="debug_agent"
2020-09-03 23:11:22 +0900 [info]: #0 starting fluentd worker pid=2905 ppid=2886 worker=0
2020-09-03 23:11:22 +0900 [info]: #0 [input_debug_agent] listening dRuby uri="druby://127.0.0.1:24230" object="Fluent::Engine" worker=0
2020-09-03 23:11:22 +0900 [info]: #0 [input_forward] listening port port=24224 bind="0.0.0.0"
2020-09-03 23:11:22 +0900 [info]: #0 fluentd worker is now running worker=0
2020-09-03 23:11:27 +0900 [info]: Received graceful stop
2020-09-03 23:11:28 +0900 [info]: #0 fluentd worker is now stopping worker=0
2020-09-03 23:11:28 +0900 [info]: #0 shutting down fluentd worker worker=0
2020-09-03 23:11:28 +0900 [info]: #0 shutting down input plugin type=:forward plugin_id="input_forward"
2020-09-03 23:11:28 +0900 [info]: #0 shutting down input plugin type=:http plugin_id="input_http"
2020-09-03 23:11:28 +0900 [info]: #0 shutting down input plugin type=:debug_agent plugin_id="input_debug_agent"
2020-09-03 23:11:28 +0900 [info]: #0 shutting down output plugin type=:tdlog plugin_id="output_td"
2020-09-03 23:11:28 +0900 [info]: #0 shutting down output plugin type=:stdout plugin_id="output_stdout"
2020-09-03 23:11:29 +0900 [info]: Worker 0 finished with status 0
2020-09-04 11:03:15 +0900 [info]: parsing config file is succeeded path="/etc/td-agent/td-agent.conf"
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-elasticsearch' version '4.1.1'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-kafka' version '0.14.1'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-prometheus' version '1.8.2'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-prometheus_pushgateway' version '0.0.2'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-record-modifier' version '2.1.0'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '2.3.0'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-s3' version '1.4.0'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-systemd' version '1.0.2'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-td' version '1.1.0'
2020-09-04 11:03:15 +0900 [info]: gem 'fluent-plugin-webhdfs' version '1.2.5'
2020-09-04 11:03:15 +0900 [info]: gem 'fluentd' version '1.11.2'
2020-09-04 11:03:16 +0900 [warn]: [output_td] Use different plugin for secondary. Check the plugin works with primary like secondary_file primary="Fluent::Plugin::TreasureDataLogOutput" secondary="Fluent::Plugin::FileOutput"
2020-09-04 11:03:17 +0900 [info]: using configuration file: <ROOT>
  <match td.*.*>
    @type tdlog
    @id output_td
    apikey xxxxxx
    auto_create_table 
    <buffer>
      @type "file"
      path "/var/log/td-agent/buffer/td"
    </buffer>
    <secondary>
      @type "file"
      path "/var/log/td-agent/failed_records"
      <buffer time>
        path /var/log/td-agent/failed_records
      </buffer>
    </secondary>
  </match>
  <match debug.**>
    @type stdout
    @id output_stdout
  </match>
  <source>
    @type forward
    @id input_forward
  </source>
  <source>
    @type http
    @id input_http
    port 8888
  </source>
  <source>
    @type debug_agent
    @id input_debug_agent
    bind "127.0.0.1"
    port 24230
  </source>
</ROOT>
2020-09-04 11:03:17 +0900 [info]: starting fluentd-1.11.2 pid=10238 ruby="2.7.1"
2020-09-04 11:03:17 +0900 [info]: spawn command to main:  cmdline=["/opt/td-agent/bin/ruby", "-Eascii-8bit:ascii-8bit", "/opt/td-agent/usr/sbin/td-agent", "--log", "/var/log/td-agent/td-agent.log", "--use-v1-config", "--under-supervisor"]
2020-09-04 11:03:20 +0900 [info]: adding match pattern="td.*.*" type="tdlog"
2020-09-04 11:03:20 +0900 [warn]: #0 [output_td] Use different plugin for secondary. Check the plugin works with primary like secondary_file primary="Fluent::Plugin::TreasureDataLogOutput" secondary="Fluent::Plugin::FileOutput"
2020-09-04 11:03:20 +0900 [info]: adding match pattern="debug.**" type="stdout"
2020-09-04 11:03:21 +0900 [info]: adding source type="forward"
2020-09-04 11:03:21 +0900 [info]: adding source type="http"
2020-09-04 11:03:21 +0900 [info]: adding source type="debug_agent"
2020-09-04 11:03:21 +0900 [info]: #0 starting fluentd worker pid=10252 ppid=10238 worker=0
2020-09-04 11:03:22 +0900 [info]: #0 [input_debug_agent] listening dRuby uri="druby://127.0.0.1:24230" object="Fluent::Engine" worker=0
2020-09-04 11:03:22 +0900 [info]: #0 [input_forward] listening port port=24224 bind="0.0.0.0"
2020-09-04 11:03:22 +0900 [info]: #0 fluentd worker is now running worker=0
```

File: Rakefile [hunk 7]
```diff
@@ -51,6 +51,7 @@ end
   ["msi:build",        "Build MSI package (alias for msi:dockerbuild)"],
   ["msi:selfbuild",    "Build MSI package without using Docker"],
   ["msi:dockerbuild",  "Build MSI package by Docker"],
+  ["dmg:selfbuild",    "Build macOS package without using Docker"],
```
Comment (ashie) [inline Rakefile:54]: If we use `dmg:selfbuild` instead of `macos:selfbuild`, directory name should be also `dmg` to unify naming rule.

Comment (cosmo0920) [reply Rakefile:54]: Sure. Will do.

**Comment by cosmo0920** [comment]
@ashie @kenhys Any concerns?

**Comment by ashie** [comment]
It's not this PR's problem but I've noticed that `bundle package` still requires `sudo` if `BUNDLE_PATH` isn't set, it attempts to install fluentd gem:

```
$ rake dmg:selfbuild
cd td-agent
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby -S rake dmg:selfbuild
Following files may not be writable, so sudo is needed:
  /Library/Ruby/Gems/2.6.0
  /Library/Ruby/Gems/2.6.0/bin
  /Library/Ruby/Gems/2.6.0/build_info
  /Library/Ruby/Gems/2.6.0/cache
  /Library/Ruby/Gems/2.6.0/doc
  /Library/Ruby/Gems/2.6.0/extensions
  /Library/Ruby/Gems/2.6.0/gems
  /Library/Ruby/Gems/2.6.0/specifications
  /usr/local/bin
Your Gemfile contains path and git dependencies. If you want to cache them as well, please pass the --all flag. This will be the default on Bundler 3.0.


Your user account isn't allowed to install to the system RubyGems.
  You can cancel this installation and run:

      bundle install --path vendor/bundle

  to install the gems into ./vendor/bundle/, or you can enter your password
  and install the bundled gems to RubyGems using sudo.

  Password: 
```

**Review by ashie — APPROVED** [review]

**Comment by cosmo0920** [comment]
Thanks!

**Comment by cosmo0920** [comment]
> It's not this PR's problem but I've noticed that `bundle package` still requires `sudo` if `BUNDLE_PATH` isn't set, it attempts to install fluentd gem:

Ah! I'm using rbenv's Ruby. So, I didn't met this circumstances....

**Comment by ashie** [comment]
> It's not this PR's problem but I've noticed that bundle package still requires sudo if BUNDLE_PATH isn't set, it attempts to install fluentd gem:

#194

**Pull request merged**